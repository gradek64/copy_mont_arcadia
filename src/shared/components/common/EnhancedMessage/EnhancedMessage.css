/** @define EnhancedMessage; weak */

.EnhancedMessage {
  box-sizing: border-box;
  overflow: hidden;

  &-wrapper {
    background-color: #fff;
    border: 1px solid $dk-gray;
    display: flex;
    flex-direction: column;
    margin: 0 0 5px;
  }

  &-section {
    display: flex;
    flex-direction: row;
    padding: 10px;
  }

  &-icon {
    background-image: $icon-info;
    background-position: center;
    background-repeat: no-repeat;
    background-size: contain;
    margin: 0 10px 0 0;
    width: 20px;
  }

  &-content {
    margin: 0;
  }

  &-header {
    color: $dk-gray;
    font-size: 12px;
    margin: 0;
    letter-spacing: 0;
  }

  &-message {
    color: $dk-gray;
    display: block;
    font-size: 14px;
  }

  &.is-error {
    & .EnhancedMessage-wrapper {
      border: 1px solid #f00;
    }

    & .EnhancedMessage-icon {
      background-image: $icon-info-error;
    }

    & .EnhancedMessage-content {
      color: $error-color;
    }

    & .EnhancedMessage-header {
      color: $error-color;
    }

    & .EnhancedMessage-message {
      color: $error-color;
    }
  }

}

/* postcss-bem-linter: ignore */
.EnhancedMessage.isFromFilter {
  margin: 20px;
}

/* postcss-bem-linter: ignore */
.EnhancedMessage-section.isFromFilter {
  justify-content: center;
}

@media $min-tablet {
  /* postcss-bem-linter: ignore */
  .EnhancedMessage-wrapper.isFromFilter {
    padding: 10px;
  }

  /* postcss-bem-linter: ignore */
  .Button.isFromFilter {
    width: 50%;
    margin: 10px auto;
  }
}

/* postcss-bem-linter: ignore */
.EnhancedMessage-enter {
  max-height: 0;
  opacity: 0.01;
}

/* postcss-bem-linter: ignore */
.EnhancedMessage-enter-active {
  max-height: 500px;
  opacity: 1;
  transition: all 300ms ease-in-out;
}

/* postcss-bem-linter: ignore */
.EnhancedMessage-exit {
  max-height: 500px;
  opacity: 1;
}

/* postcss-bem-linter: ignore */
.EnhancedMessage-exit-active {
  max-height: 0;
  opacity: 0.01;
  transition: all 300ms ease-in-out;
}
