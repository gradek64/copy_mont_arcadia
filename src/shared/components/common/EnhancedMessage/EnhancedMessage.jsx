import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import { CSSTransition } from 'react-transition-group'

/*
  @param header - optional bold header
  @param message - optional string or array of strings
  @param duration - length of time in ms for message to persist after transition in
  @param isError - true to display with error message styling
  @param isVisible - true to display immediately on render, false to delay display (can be triggered via redux action)
  @param showOnce - false to display permanently, true to display only once and then remove if onTransitionComplete is provided
  @param onTransitionComplete - optional callback, triggered after animation completes if duration value provided, or on unmount if showOnce=true
  @param renderChildrenOn - optional location to render children (top(default), bottom)
  @param isFromFilter - optional flag adds relevant filter classes for styling
*/
class EnhancedMessage extends Component {
  constructor(props) {
    super(props)

    this.state = { isVisible: false }
  }

  static propTypes = {
    id: PropTypes.string,
    header: PropTypes.string,
    message: PropTypes.oneOfType([PropTypes.string, PropTypes.array]),
    duration: PropTypes.oneOfType([PropTypes.bool, PropTypes.number]),
    isError: PropTypes.bool,
    isVisible: PropTypes.bool,
    showOnce: PropTypes.bool,
    onTransitionComplete: PropTypes.func,
    renderChildrenOn: PropTypes.string,
    isFromFilter: PropTypes.bool,
    children: PropTypes.node,
  }

  static defaultProps = {
    duration: false,
    isError: false,
    isVisible: true,
    showOnce: false,
    onTransitionComplete: () => {},
    renderChildrenOn: 'top',
    isFromFilter: false,
  }

  componentDidMount() {
    const { isVisible } = this.props
    this.setDisplayState(isVisible)
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    if (this.props.isVisible !== nextProps.isVisible)
      this.setDisplayState(nextProps.isVisible)
  }

  componentWillUnmount() {
    const { showOnce } = this.props
    if (showOnce) this.handleOnExited()
  }

  shouldComponentUpdate(nextProps, nextState) {
    // only update if local state changes to avoid synchronisation issues with animations
    return this.state.isVisible !== nextState.isVisible
  }

  setDisplayState = (isVisible) => {
    this.setState({ isVisible })
  }

  handleOnEntered = () => {
    const { duration } = this.props
    if (duration) this.setDisplayState(false)
  }

  handleOnExited = () => {
    const { id, onTransitionComplete } = this.props
    if (id) onTransitionComplete(id)
  }

  renderMessage = () => {
    const {
      children,
      header,
      message,
      isError,
      renderChildrenOn,
      isFromFilter,
    } = this.props

    const messageClass = classnames('EnhancedMessage', {
      'is-error': isError,
      isFromFilter,
    })

    return message || children ? (
      <div className={messageClass}>
        <div
          className={`EnhancedMessage-wrapper ${
            isFromFilter ? 'isFromFilter' : ''
          }`}
        >
          {renderChildrenOn === 'top' && children}
          {message && (
            <div
              className={`EnhancedMessage-section ${
                isFromFilter ? 'isFromFilter' : ''
              }`}
            >
              <span className="EnhancedMessage-icon" />
              <p className="EnhancedMessage-content">
                {header && <h5 className="EnhancedMessage-header">{header}</h5>}
                {message instanceof Array ? (
                  message.map((msg) => (
                    <span className="EnhancedMessage-message">{msg}</span>
                  ))
                ) : (
                  <span className="EnhancedMessage-message">{message}</span>
                )}
              </p>
            </div>
          )}
          {renderChildrenOn === 'bottom' && children}
        </div>
      </div>
    ) : null
  }

  render() {
    const { duration, showOnce } = this.props
    const { isVisible } = this.state

    return duration ? (
      <CSSTransition
        in={isVisible}
        timeout={{ enter: duration, exit: 500 }}
        classNames="EnhancedMessage"
        onEntered={this.handleOnEntered}
        onExited={this.handleOnExited}
        unmountOnExit={showOnce}
      >
        {this.renderMessage()}
      </CSSTransition>
    ) : (
      this.renderMessage()
    )
  }
}

export default EnhancedMessage
