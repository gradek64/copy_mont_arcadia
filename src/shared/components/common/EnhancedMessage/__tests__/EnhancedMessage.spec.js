import testComponentHelper from 'test/unit/helpers/test-component'
import EnhancedMessage from '../EnhancedMessage'
import { CSSTransition } from 'react-transition-group'

describe('<EnhancedMessage />', () => {
  const initialProps = {
    onTransitionComplete: jest.fn(),
  }
  const renderComponent = testComponentHelper(EnhancedMessage)

  describe('@renders', () => {
    it('renders in default state', () => {
      const component = renderComponent(initialProps)
      expect(component.getTree()).toMatchSnapshot()
    })
    it('renders with optional header', () => {
      const { wrapper } = renderComponent({
        ...initialProps,
        header: 'this is a header',
        message: 'this is a message',
      })
      expect(wrapper.find('.EnhancedMessage-header')).toHaveLength(1)
    })
    it('renders if message is a string', () => {
      const { wrapper } = renderComponent({
        ...initialProps,
        message: 'this is a message',
      })
      expect(wrapper.find('.EnhancedMessage-message')).toHaveLength(1)
    })
    it('renders if array of messages', () => {
      const { wrapper } = renderComponent({
        ...initialProps,
        message: ['this is a message', 'this is another message'],
      })
      expect(wrapper.find('.EnhancedMessage-message')).toHaveLength(2)
    })
    it('renders if children passed (default)', () => {
      const props = {
        ...initialProps,
        message: 'hello World',
        children: '<div>Children</div>',
      }
      const component = renderComponent(props)
      expect(component.getTree()).toMatchSnapshot()
    })
    it('renders if children passed ("bottom")', () => {
      const props = {
        ...initialProps,
        message: 'hello World',
        children: '<div>Children</div>',
        renderChildrenOn: 'bottom',
      }
      const component = renderComponent(props)
      expect(component.getTree()).toMatchSnapshot()
    })
    it('does not render if no message or children', () => {
      const { wrapper } = renderComponent(initialProps)
      expect(wrapper.find('.EnhancedMessage').exists()).toEqual(false)
    })
    it('renders with error class if isError is true', () => {
      const { wrapper } = renderComponent({
        ...initialProps,
        message: 'this is an error message',
        isError: true,
      })
      expect(wrapper.find('.EnhancedMessage.is-error')).toHaveLength(1)
    })
    it('renders with animation capability if duration provided', () => {
      const { wrapper } = renderComponent({
        ...initialProps,
        message: 'this is a message',
        duration: 1000,
      })
      expect(wrapper.find(CSSTransition)).toHaveLength(1)
      expect(wrapper.find('.EnhancedMessage')).toHaveLength(1)
    })
    it('renders without animation capability if duration not provided', () => {
      const { wrapper } = renderComponent({
        ...initialProps,
        message: 'this is a message',
      })
      expect(wrapper.find(CSSTransition).exists()).toEqual(false)
      expect(wrapper.find('.EnhancedMessage')).toHaveLength(1)
    })
  })

  describe('@lifecycle', () => {
    beforeEach(() => jest.resetAllMocks())

    describe('constructor', () => {
      it('sets initial state for message visibility', () => {
        const { instance } = renderComponent(initialProps)
        expect(instance.state).toEqual({ isVisible: false })
      })
    })

    describe('componentDidMount', () => {
      it('sets the message visibility state to true', () => {
        const { instance } = renderComponent(initialProps)
        const setDisplayStateSpy = jest.spyOn(instance, 'setDisplayState')
        instance.componentDidMount()
        expect(setDisplayStateSpy).toHaveBeenCalledTimes(1)
        expect(setDisplayStateSpy).toHaveBeenCalledWith(true)
      })
    })

    describe('UNSAFE_componentWillReceiveProps', () => {
      it('calls `setDisplayState` with `nextProps.isVisible` if changed', () => {
        const { instance } = renderComponent({
          ...initialProps,
          isVisible: false,
        })
        const setDisplayStateSpy = jest.spyOn(instance, 'setDisplayState')
        instance.state = { isVisible: false }
        instance.UNSAFE_componentWillReceiveProps({ isVisible: true })
        expect(setDisplayStateSpy).toHaveBeenCalled()
        expect(setDisplayStateSpy).toHaveBeenCalledWith(true)
      })
    })

    describe('componentWillUnmount', () => {
      it('calls instance method to trigger callback if showOnce is true', () => {
        const { instance } = renderComponent({
          ...initialProps,
          showOnce: true,
        })
        const setHandleOnExitedSpy = jest.spyOn(instance, 'handleOnExited')
        instance.componentWillUnmount()
        expect(setHandleOnExitedSpy).toHaveBeenCalledTimes(1)
      })
    })

    describe('shouldComponentUpdate', () => {
      it('returns true if state changes', () => {
        const { instance } = renderComponent(initialProps)
        instance.state = { isVisible: false }
        expect(instance.shouldComponentUpdate({}, { isVisible: true })).toEqual(
          true
        )
      })

      it('returns false if state does not change', () => {
        const { instance } = renderComponent({
          ...initialProps,
        })
        instance.state = { isVisible: false }
        expect(
          instance.shouldComponentUpdate({}, { isVisible: false })
        ).toEqual(false)
      })
    })
  })

  describe('instance methods', () => {
    describe('setDisplayState', () => {
      it('updates the message visibility state', () => {
        const { instance } = renderComponent(initialProps)
        const isVisible = true
        const setStateSpy = jest.spyOn(instance, 'setState')
        instance.setDisplayState(isVisible)
        expect(setStateSpy).toHaveBeenCalledTimes(1)
        expect(setStateSpy).toHaveBeenCalledWith({ isVisible })
      })
    })

    describe('handleOnEntered', () => {
      it('sets the message visibility state to false after `enter` transition if transient message', () => {
        const { instance } = renderComponent({
          ...initialProps,
          duration: 1000,
        })
        const setDisplayStateSpy = jest.spyOn(instance, 'setDisplayState')
        instance.handleOnEntered()
        expect(setDisplayStateSpy).toHaveBeenCalledTimes(1)
        expect(setDisplayStateSpy).toHaveBeenCalledWith(false)
      })
      it('does not change message visibility state to false after `enter` transition if permanent message', () => {
        const { instance } = renderComponent({
          ...initialProps,
          duration: false,
        })
        const setDisplayStateSpy = jest.spyOn(instance, 'setDisplayState')
        instance.handleOnEntered()
        expect(setDisplayStateSpy).not.toHaveBeenCalled()
      })
    })

    describe('handleOnExited', () => {
      it('triggers callback method after entire transition completes', () => {
        const id = '1234'
        const { instance } = renderComponent({
          ...initialProps,
          id,
        })
        instance.handleOnExited(id)
        expect(initialProps.onTransitionComplete).toHaveBeenCalledTimes(1)
        expect(initialProps.onTransitionComplete).toHaveBeenCalledWith(id)
      })
    })
  })
})
