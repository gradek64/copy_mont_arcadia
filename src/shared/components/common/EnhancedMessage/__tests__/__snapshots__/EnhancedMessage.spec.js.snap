// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EnhancedMessage /> @renders renders if children passed ("bottom") 1`] = `
<div
  className="EnhancedMessage"
>
  <div
    className="EnhancedMessage-wrapper "
  >
    <div
      className="EnhancedMessage-section "
    >
      <span
        className="EnhancedMessage-icon"
      />
      <p
        className="EnhancedMessage-content"
      >
        <span
          className="EnhancedMessage-message"
        >
          hello World
        </span>
      </p>
    </div>
    &lt;div&gt;Children&lt;/div&gt;
  </div>
</div>
`;

exports[`<EnhancedMessage /> @renders renders if children passed (default) 1`] = `
<div
  className="EnhancedMessage"
>
  <div
    className="EnhancedMessage-wrapper "
  >
    &lt;div&gt;Children&lt;/div&gt;
    <div
      className="EnhancedMessage-section "
    >
      <span
        className="EnhancedMessage-icon"
      />
      <p
        className="EnhancedMessage-content"
      >
        <span
          className="EnhancedMessage-message"
        >
          hello World
        </span>
      </p>
    </div>
  </div>
</div>
`;

exports[`<EnhancedMessage /> @renders renders in default state 1`] = `""`;
